error when compiling on macOS

Is catfs supposted to run on macOS?

Errors below occured when I try to install it with cargo.

```
error[E0425]: cannot find function `fdopendir` in module `libc`
  --> src/catfs/rlibc.rs:56:29
   |
56 |     let dh = unsafe { libc::fdopendir(fd) };
   |                             ^^^^^^^^^ did you mean `opendir`?
help: possible candidate is found in another module, you can import it into scope
   |
6  | use catfs::rlibc::fdopendir;
   |

error[E0425]: cannot find function `pipe2` in module `libc`
   --> src/catfs/rlibc.rs:178:30
    |
178 |     let res = unsafe { libc::pipe2(p.as_mut_ptr(), libc::O_CLOEXEC) };
    |                              ^^^^^ did you mean `pipe`?

error[E0425]: cannot find function `splice` in module `libc`
   --> src/catfs/rlibc.rs:207:30
    |
207 |     let res = unsafe { libc::splice(fd, off_from_ptr, other, off_to_ptr, len, 0) };
    |                              ^^^^^^ not found in `libc`
help: possible candidate is found in another module, you can import it into scope
    |
6   | use catfs::rlibc::splice;
    |

error[E0425]: cannot find value `AT_EMPTY_PATH` in module `libc`
   --> src/catfs/rlibc.rs:279:66
    |
279 |     let res = unsafe { libc::fstatat(dir, s.as_ptr(), stp, libc::AT_EMPTY_PATH) };
    |                                                                  ^^^^^^^^^^^^^ not found in `libc`

error[E0425]: cannot find function `utimensat` in module `libc`
   --> src/catfs/rlibc.rs:342:30
    |
342 |     let res = unsafe { libc::utimensat(dir, s.as_ptr(), times.as_mut_ptr(), flags as i32) };
    |                              ^^^^^^^^^ did you mean `utimes`?
help: possible candidate is found in another module, you can import it into scope
    |
6   | use catfs::rlibc::utimensat;
    |

error[E0425]: cannot find function `posix_fallocate` in module `libc`
   --> src/catfs/rlibc.rs:435:34
    |
435 |         let res = unsafe { libc::posix_fallocate(self.fd, offset as i64, len as i64) };
    |                                  ^^^^^^^^^^^^^^^ not found in `libc`

error[E0308]: mismatched types
   --> src/catfs/mod.rs:195:21
    |
195 |                     st.f_blocks,
    |                     ^^^^^^^^^^^ expected u64, found u32
help: you can cast an `u32` to `u64`, which will zero-extend the source value
    |
195 |                     st.f_blocks.into(),
    |                     ^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> src/catfs/mod.rs:196:21
    |
196 |                     st.f_bfree,
    |                     ^^^^^^^^^^ expected u64, found u32
help: you can cast an `u32` to `u64`, which will zero-extend the source value
    |
196 |                     st.f_bfree.into(),
    |                     ^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> src/catfs/mod.rs:197:21
    |
197 |                     st.f_bavail,
    |                     ^^^^^^^^^^^ expected u64, found u32
help: you can cast an `u32` to `u64`, which will zero-extend the source value
    |
197 |                     st.f_bavail.into(),
    |                     ^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> src/catfs/mod.rs:198:21
    |
198 |                     st.f_files,
    |                     ^^^^^^^^^^ expected u64, found u32
help: you can cast an `u32` to `u64`, which will zero-extend the source value
    |
198 |                     st.f_files.into(),
    |                     ^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> src/catfs/mod.rs:199:21
    |
199 |                     st.f_ffree,
    |                     ^^^^^^^^^^ expected u64, found u32
help: you can cast an `u32` to `u64`, which will zero-extend the source value
    |
199 |                     st.f_ffree.into(),
    |                     ^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> src/catfs/file.rs:598:56
    |
598 |         self.src_file = File::openat(dir, path, flags, mode)?;
    |                                                        ^^^^ expected u32, found u16
help: you can cast an `u16` to `u32`, which will zero-extend the source value
    |
598 |         self.src_file = File::openat(dir, path, flags, mode.into())?;
    |                                                        ^^^^^^^^^^^

error[E0560]: struct `catfs::libc::dirent` has no field named `d_off`
  --> src/catfs/rlibc.rs:89:17
   |
89 |                 d_off: 0,
   |                 ^^^^^^ `catfs::libc::dirent` does not have this field
   |
   = note: available fields are: `d_ino`, `d_seekoff`, `d_reclen`, `d_namlen`, `d_type`, `d_name`

error[E0308]: mismatched types
  --> src/catfs/rlibc.rs:92:25
   |
92 |                 d_name: [0i8; 256], // FIXME: don't hardcode 256
   |                         ^^^^^^^^^^ expected an array with a fixed size of 1024 elements, found one with 256 elements
   |
   = note: expected type `[i8; 1024]`
              found type `[i8; 256]`

error[E0609]: no field `d_off` on type `catfs::libc::dirent`
   --> src/catfs/rlibc.rs:120:24
    |
120 |         return self.en.d_off;
    |                        ^^^^^ unknown field
    |
    = note: available fields are: `d_ino`, `d_seekoff`, `d_reclen`, `d_namlen`, `d_type`, `d_name`

error[E0308]: mismatched types
   --> src/catfs/rlibc.rs:158:48
    |
158 |     let res = unsafe { libc::mkdir(s.as_ptr(), mode) };
    |                                                ^^^^ expected u16, found u32

error[E0308]: mismatched types
   --> src/catfs/rlibc.rs:168:55
    |
168 |     let res = unsafe { libc::mkdirat(dir, s.as_ptr(), mode) };
    |                                                       ^^^^ expected u16, found u32

error[E0308]: mismatched types
   --> src/catfs/rlibc.rs:352:56
    |
352 |     let res = unsafe { libc::fchmodat(dir, s.as_ptr(), mode, flags as i32) };
    |                                                        ^^^^ expected u16, found u32

error[E0308]: mismatched types
   --> src/catfs/rlibc.rs:459:50
    |
459 |         let res = unsafe { libc::fchmod(self.fd, mode) };
    |                                                  ^^^^ expected u16, found u32

error[E0308]: mismatched types
  --> src/catfs/inode.rs:39:15
   |
39 |     match t & libc::S_IFMT {
   |               ^^^^^^^^^^^^ expected u32, found u16

error[E0308]: mismatched types
  --> src/catfs/inode.rs:40:9
   |
40 |         libc::S_IFLNK => fuse::FileType::Symlink,
   |         ^^^^^^^^^^^^^ expected u32, found u16
   |
   = note: expected type `u32`
              found type `u16`

error[E0308]: mismatched types
  --> src/catfs/inode.rs:41:9
   |
41 |         libc::S_IFREG => fuse::FileType::RegularFile,
   |         ^^^^^^^^^^^^^ expected u32, found u16
   |
   = note: expected type `u32`
              found type `u16`

error[E0308]: mismatched types
  --> src/catfs/inode.rs:42:9
   |
42 |         libc::S_IFBLK => fuse::FileType::BlockDevice,
   |         ^^^^^^^^^^^^^ expected u32, found u16
   |
   = note: expected type `u32`
              found type `u16`

error[E0308]: mismatched types
  --> src/catfs/inode.rs:43:9
   |
43 |         libc::S_IFDIR => fuse::FileType::Directory,
   |         ^^^^^^^^^^^^^ expected u32, found u16
   |
   = note: expected type `u32`
              found type `u16`

error[E0308]: mismatched types
  --> src/catfs/inode.rs:44:9
   |
44 |         libc::S_IFCHR => fuse::FileType::CharDevice,
   |         ^^^^^^^^^^^^^ expected u32, found u16
   |
   = note: expected type `u32`
              found type `u16`

error[E0308]: mismatched types
  --> src/catfs/inode.rs:45:9
   |
45 |         libc::S_IFIFO => fuse::FileType::NamedPipe,
   |         ^^^^^^^^^^^^^ expected u32, found u16
   |
   = note: expected type `u32`
              found type `u16`

error[E0277]: no implementation for `u32 & u16`
  --> src/catfs/inode.rs:39:13
   |
39 |     match t & libc::S_IFMT {
   |             ^ no implementation for `u32 & u16`
   |
   = help: the trait `std::ops::BitAnd<u16>` is not implemented for `u32`

error[E0308]: mismatched types
   --> src/catfs/inode.rs:131:31
    |
131 |             kind: to_filetype(st.st_mode),
    |                               ^^^^^^^^^^ expected u32, found u16
help: you can cast an `u16` to `u32`, which will zero-extend the source value
    |
131 |             kind: to_filetype(st.st_mode.into()),
    |                               ^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> src/evicter/mod.rs:104:50
    |
104 |         DiskSpace::Percent(p) => ((st.f_blocks * st.f_frsize) as f64 * p / 100.0) as u64,
    |                                                  ^^^^^^^^^^^ expected u32, found u64

error[E0277]: cannot multiply `u64` to `u32`
   --> src/evicter/mod.rs:104:48
    |
104 |         DiskSpace::Percent(p) => ((st.f_blocks * st.f_frsize) as f64 * p / 100.0) as u64,
    |                                                ^ no implementation for `u32 * u64`
    |
    = help: the trait `std::ops::Mul<u64>` is not implemented for `u32`

error[E0308]: mismatched types
   --> src/evicter/mod.rs:108:37
    |
108 |     let x = desired - (st.f_bfree * st.f_frsize) as i64;
    |                                     ^^^^^^^^^^^ expected u32, found u64

error[E0277]: cannot multiply `u64` to `u32`
   --> src/evicter/mod.rs:108:35
    |
108 |     let x = desired - (st.f_bfree * st.f_frsize) as i64;
    |                                   ^ no implementation for `u32 * u64`
    |
    = help: the trait `std::ops::Mul<u64>` is not implemented for `u32`

error: aborting due to 32 previous errors

error: failed to compile `catfs v0.8.0`, intermediate artifacts can be found at `/var/folders/8j/gcq2wfcj41jbn45x071shl4h0000gn/T/cargo-install.75RhyeDLPNLv`
```

--- a/src/catfs/file.rs
+++ b/src/catfs/file.rs
@@ -270,7 +270,8 @@ impl Handle {
             )?;
         } else {
             if let Err(e) = self.cache_file.remove_xattr("user.catfs.src_chksum") {
-                if e.raw_os_error().unwrap() != libc::ENODATA {
+                let my_errno = e.raw_os_error().unwrap();
+                if my_errno != libc::ENODATA && my_errno != libc::ENOATTR {
                     return Err(RError::from(e));
                 }
             }
@@ -615,6 +616,7 @@ impl Handle {
         return Ok(offset);
     }
 
+    #[cfg(not(target_os = "macos"))]
     fn copy_splice(&self, rh: &File, wh: &File) -> error::Result<i64> {
         let (pin, pout) = rlibc::pipe()?;
 
@@ -644,6 +646,11 @@ impl Handle {
         }
 
         return Ok(offset);
+    }
+
+    #[cfg(target_os = "macos")]
+    fn copy_splice(&self, rh: &File, wh: &File) -> error::Result<i64> {
+        self.copy_user(rh, wh)
     }
 
     fn copy(&self, to_cache: bool, disable_splice: bool) -> error::Result<()> {

--- a/src/catfs/rlibc.rs
+++ b/src/catfs/rlibc.rs
@@ -9,6 +9,7 @@ use std::os::unix::ffi::{OsStrExt, OsStringExt};
 use std::io;
 use std::mem;
 use std::path::Path;
+#[cfg(not(target_os = "macos"))]
 use std::ptr;
 use std::os::unix::io::AsRawFd;
 use std::os::unix::io::RawFd;
@@ -17,6 +18,11 @@ use std::os::unix::fs::FileExt;
 use self::fuse::FileType;
 use self::time::Timespec;
 use self::xattr::FileExt as XattrFileExt;
+
+#[cfg(not(target_os = "macos"))]
+use self::libc::{fstat64, fstatvfs64, ftruncate64, open64, openat64, pread64, pwrite64, stat64, statvfs64};
+#[cfg(target_os = "macos")]
+use self::libc::{fstat as fstat64, fstatvfs as fstatvfs64, ftruncate as ftruncate64, open as open64, openat as openat64, pread as pread64, pwrite as pwrite64, stat as stat64, statvfs as statvfs64};
 
 use catfs::error;
 use catfs::error::RError;
@@ -102,6 +108,7 @@ pub struct Dirent {
 }
 
 impl Default for Dirent {
+    #[cfg(not(target_os = "macos"))]
     fn default() -> Dirent {
         return Dirent {
             en: libc::dirent {
@@ -110,6 +117,19 @@ impl Default for Dirent {
                 d_reclen: 0,
                 d_type: libc::DT_REG,
                 d_name: [0i8 as libc::c_char; 256], // FIXME: don't hardcode 256
+            },
+        };
+    }
+    #[cfg(target_os = "macos")]
+    fn default() -> Dirent {
+        return Dirent {
+            en: libc::dirent {
+                d_ino: 0,
+                d_seekoff: 0,
+                d_reclen: 0,
+                d_type: libc::DT_REG,
+                d_name: [0i8; 1024], // FIXME: don't hardcode 1024
+                d_namlen: 0,
             },
         };
     }
@@ -137,7 +157,10 @@ impl Dirent {
         return self.en.d_ino as u64;
     }
     pub fn off(&self) -> i64 {
+        #[cfg(not(target_os = "macos"))]
         return self.en.d_off as i64;
+        #[cfg(target_os = "macos")]
+        return self.en.d_seekoff as i64;
     }
     pub fn kind(&self) -> fuse::FileType {
         match self.en.d_type {
@@ -193,6 +216,7 @@ pub fn mkdirat(dir: RawFd, path: &dyn AsRef<Path>, mode: libc::mode_t) -> io::Re
     }
 }
 
+#[cfg(not(target_os = "macos"))]
 pub fn pipe() -> io::Result<(libc::c_int, libc::c_int)> {
     let mut p = [0; 2];
     let res = unsafe { libc::pipe2(p.as_mut_ptr(), libc::O_CLOEXEC) };
@@ -203,6 +227,7 @@ pub fn pipe() -> io::Result<(libc::c_int, libc::c_int)> {
     }
 }
 
+#[cfg(not(target_os = "macos"))]
 pub fn splice(
     fd: libc::c_int,
     off_self: i64,
@@ -280,10 +305,10 @@ pub fn renameat(dir: RawFd, path: &dyn AsRef<Path>, newpath: &dyn AsRef<Path>) -
     }
 }
 
-pub fn fstat(fd: libc::c_int) -> io::Result<libc::stat64> {
-    let mut st: libc::stat64 = unsafe { mem::zeroed() };
+pub fn fstat(fd: libc::c_int) -> io::Result<stat64> {
+    let mut st: stat64 = unsafe { mem::zeroed() };
 
-    let res = unsafe { libc::fstat64(fd, (&mut st) as *mut libc::stat64) };
+    let res = unsafe { fstat64(fd, (&mut st) as *mut stat64) };
     if res < 0 {
         return Err(io::Error::last_os_error());
     } else {
@@ -291,12 +316,15 @@ pub fn fstat(fd: libc::c_int) -> io::Result<libc::stat64> {
     }
 }
 
-pub fn fstatat(dir: RawFd, path: &dyn AsRef<Path>) -> io::Result<libc::stat64> {
-    let mut st: libc::stat64 = unsafe { mem::zeroed() };
-    let stp = (&mut st) as *mut libc::stat64;
+pub fn fstatat(dir: RawFd, path: &dyn AsRef<Path>) -> io::Result<stat64> {
+    let mut st: stat64 = unsafe { mem::zeroed() };
+    let stp = (&mut st) as *mut stat64;
     let s = to_cstring(path);
 
+    #[cfg(not(target_os = "macos"))]
     let res = unsafe { libc::fstatat64(dir, s.as_ptr(), stp, libc::AT_EMPTY_PATH) };
+    #[cfg(target_os = "macos")]
+    let res = unsafe { libc::fstatat(dir, s.as_ptr(), stp, 0) };
 
     if res < 0 {
         return Err(io::Error::last_os_error());
@@ -305,9 +333,9 @@ pub fn fstatat(dir: RawFd, path: &dyn AsRef<Path>) -> io::Result<libc::stat64> {
     }
 }
 
-pub fn fstatvfs(fd: RawFd) -> io::Result<libc::statvfs64> {
-    let mut st: libc::statvfs64 = unsafe { mem::zeroed() };
-    let res = unsafe { libc::fstatvfs64(fd, &mut st as *mut libc::statvfs64) };
+pub fn fstatvfs(fd: RawFd) -> io::Result<statvfs64> {
+    let mut st: statvfs64 = unsafe { mem::zeroed() };
+    let res = unsafe { fstatvfs64(fd, &mut st as *mut statvfs64) };
     if res < 0 {
         return Err(io::Error::last_os_error());
     } else {
@@ -317,7 +345,7 @@ pub fn fstatvfs(fd: RawFd) -> io::Result<libc::statvfs64> {
 
 pub fn openat(dir: RawFd, path: &dyn AsRef<Path>, flags: u32, mode: libc::mode_t) -> io::Result<RawFd> {
     let s = to_cstring(path);
-    let fd = unsafe { libc::openat64(dir, s.as_ptr(), (flags | O_CLOEXEC) as i32, mode) };
+    let fd = unsafe { openat64(dir, s.as_ptr(), (flags | O_CLOEXEC) as i32, mode as libc::c_uint) };
     if fd == -1 {
         return Err(io::Error::last_os_error());
     } else {
@@ -391,7 +419,7 @@ fn as_mut_void_ptr<T>(s: &mut [T]) -> *mut libc::c_void {
 
 pub fn open(path: &dyn AsRef<Path>, flags: u32, mode: u32) -> io::Result<RawFd> {
     let s = to_cstring(path);
-    let fd = unsafe { libc::open64(s.as_ptr(), (flags | O_CLOEXEC) as i32, mode) };
+    let fd = unsafe { open64(s.as_ptr(), (flags | O_CLOEXEC) as i32, mode as libc::c_uint) };
     if fd == -1 {
         return Err(io::Error::last_os_error());
     } else {
@@ -438,12 +466,12 @@ impl File {
         return Ok(st.st_size as u64);
     }
 
-    pub fn stat(&self) -> io::Result<libc::stat64> {
+    pub fn stat(&self) -> io::Result<stat64> {
         fstat(self.fd)
     }
 
     pub fn truncate(&self, size: u64) -> io::Result<()> {
-        let res = unsafe { libc::ftruncate64(self.fd, size as i64) };
+        let res = unsafe { ftruncate64(self.fd, size as i64) };
         if res < 0 {
             return Err(io::Error::last_os_error());
         } else {
@@ -451,6 +479,7 @@ impl File {
         }
     }
 
+    #[cfg(not(target_os = "macos"))]
     pub fn allocate(&self, offset: u64, len: u64) -> io::Result<()> {
         let res = unsafe { libc::posix_fallocate64(self.fd, offset as i64, len as i64) };
         if res == 0 {
@@ -458,6 +487,11 @@ impl File {
         } else {
             return Err(io::Error::from_raw_os_error(res));
         }
+    }
+
+    #[cfg(target_os = "macos")]
+    pub fn allocate(&self, offset: u64, len: u64) -> io::Result<()> {
+        self.truncate(offset + len)
     }
 
     #[allow(dead_code)]
@@ -486,7 +520,7 @@ impl File {
 
     pub fn read_at(&self, buf: &mut [u8], offset: i64) -> io::Result<usize> {
         let nbytes =
-            unsafe { libc::pread64(self.fd, as_mut_void_ptr(buf), buf.len(), offset) };
+            unsafe { pread64(self.fd, as_mut_void_ptr(buf), buf.len(), offset) };
         if nbytes < 0 {
             return Err(io::Error::last_os_error());
         } else {
@@ -495,7 +529,7 @@ impl File {
     }
 
     pub fn write_at(&self, buf: &[u8], offset: i64) -> io::Result<usize> {
-        let nbytes = unsafe { libc::pwrite64(self.fd, as_void_ptr(buf), buf.len(), offset) };
+        let nbytes = unsafe { pwrite64(self.fd, as_void_ptr(buf), buf.len(), offset) };
         if nbytes < 0 {
             return Err(io::Error::last_os_error());
         } else {

--- a/src/evicter/mod.rs
+++ b/src/evicter/mod.rs
@@ -24,6 +24,11 @@ use self::dir_walker::DirWalker;
 use self::itertools::Itertools;
 use self::twox_hash::XxHash;
 
+#[cfg(not(target_os = "macos"))]
+use self::libc::statvfs64;
+#[cfg(target_os = "macos")]
+use self::libc::{statvfs as statvfs64};
+
 pub struct Evicter {
     dir: RawFd,
     high_watermark: DiskSpace,
@@ -31,7 +36,7 @@ pub struct Evicter {
     scan_freq: Duration,
     hot_percent: usize, // 25 to keep most recently used 25%
     request_weight: u32,
-    statvfs: fn(RawFd) -> io::Result<libc::statvfs64>,
+    statvfs: fn(RawFd) -> io::Result<statvfs64>,
     cv: Arc<Condvar>,
     shutting_down: Arc<Mutex<bool>>,
     t: Option<JoinHandle<()>>,
@@ -99,22 +104,22 @@ impl Hasher for IdentU64Hasher {
 }
 
 // in blocks
-fn to_evict(spec: &DiskSpace, st: &libc::statvfs64) -> u64 {
+fn to_evict(spec: &DiskSpace, st: &statvfs64) -> u64 {
     let desired = match *spec {
-        DiskSpace::Percent(p) => ((st.f_blocks * st.f_frsize as u64) as f64 * p / 100.0) as u64,
+        DiskSpace::Percent(p) => ((st.f_blocks as u64 * st.f_frsize as u64) as f64 * p / 100.0) as u64,
         DiskSpace::Bytes(b) => b,
     } as i64;
 
-    let x = desired - (st.f_bfree * st.f_frsize as u64) as i64;
+    let x = desired - (st.f_bfree as u64 * st.f_frsize as u64) as i64;
     return if x > 0 { x as u64 } else { 0 };
 }
 
 impl Evicter {
-    fn should_evict(&self, st: &libc::statvfs64) -> u64 {
+    fn should_evict(&self, st: &statvfs64) -> u64 {
         return to_evict(&self.high_watermark, st);
     }
 
-    fn to_evict(&self, st: &libc::statvfs64) -> u64 {
+    fn to_evict(&self, st: &statvfs64) -> u64 {
         return to_evict(&self.low_watermark, st);
     }
 
@@ -235,7 +240,7 @@ impl Evicter {
         dir: RawFd,
         free: &DiskSpace,
         scan_freq: Duration,
-        statvfs: fn(RawFd) -> io::Result<libc::statvfs64>,
+        statvfs: fn(RawFd) -> io::Result<statvfs64>,
     ) -> Evicter {
         let mut ev = Evicter {
             dir: dir,
@@ -321,7 +326,7 @@ mod tests {
 
     #[test]
     fn to_evict_bytes() {
-        let mut st: libc::statvfs64 = unsafe { mem::zeroed() };
+        let mut st: statvfs64 = unsafe { mem::zeroed() };
         st.f_bsize = 4096;
         st.f_frsize = 4096;
         st.f_blocks = 100;
@@ -345,8 +350,8 @@ mod tests {
         let prefix = catfs::tests::copy_resources();
         let fd = rlibc::open(&prefix, rlibc::O_RDONLY, 0).unwrap();
 
-        fn fake_statvfs(_dir: RawFd) -> io::Result<libc::statvfs64> {
-            let mut st: libc::statvfs64 = unsafe { mem::zeroed() };
+        fn fake_statvfs(_dir: RawFd) -> io::Result<statvfs64> {
+            let mut st: statvfs64 = unsafe { mem::zeroed() };
             st.f_bsize = 4096;
             st.f_frsize = 4096;
             st.f_blocks = 10;
@@ -367,16 +372,16 @@ mod tests {
         let prefix = catfs::tests::copy_resources();
         let fd = rlibc::open(&prefix, rlibc::O_RDONLY, 0).unwrap();
 
-        fn fake_statvfs(dir: RawFd) -> io::Result<libc::statvfs64> {
+        fn fake_statvfs(dir: RawFd) -> io::Result<statvfs64> {
             let cache_size = count_cache_size(dir).unwrap();
 
-            let mut st: libc::statvfs64 = unsafe { mem::zeroed() };
+            let mut st: statvfs64 = unsafe { mem::zeroed() };
             st.f_bsize = 4096;
             st.f_frsize = 4096;
             st.f_blocks = 100;
             // want 1 free block at beginning. cache_size is 5 * 4K blocks so pretend
             // 94 blocks are used by other things
-            st.f_bfree = st.f_blocks - cache_size / (st.f_frsize as u64) - 94;
+            st.f_bfree = (st.f_blocks as u64 - cache_size / (st.f_frsize as u64) - 94) as libc::fsblkcnt_t;
             return Ok(st);
         }
 
@@ -403,16 +408,16 @@ mod tests {
         let prefix = catfs::tests::copy_resources();
         let fd = rlibc::open(&prefix, rlibc::O_RDONLY, 0).unwrap();
 
-        fn fake_statvfs(dir: RawFd) -> io::Result<libc::statvfs64> {
+        fn fake_statvfs(dir: RawFd) -> io::Result<statvfs64> {
             let cache_size = count_cache_size(dir).unwrap();
 
-            let mut st: libc::statvfs64 = unsafe { mem::zeroed() };
+            let mut st: statvfs64 = unsafe { mem::zeroed() };
             st.f_bsize = 4096;
             st.f_frsize = 4096;
             st.f_blocks = 100;
             // want 1 free block at beginning. cache_size is 5 * 4K blocks so pretend
             // 94 blocks are used by other things
-            st.f_bfree = st.f_blocks - cache_size / (st.f_frsize as u64) - 94;
+            st.f_bfree = (st.f_blocks as u64 - cache_size / (st.f_frsize as u64) - 94) as libc::fsblkcnt_t;
             return Ok(st);
         }
 

--- a/tests/integration_tests.rs
+++ b/tests/integration_tests.rs
@@ -228,7 +228,7 @@ unit_tests!{
         of.push(f.mnt.join("foo"));
         let status = Command::new("dd")
             .arg("if=/dev/zero").arg(of)
-            .arg("bs=1M").arg("count=10")
+            .arg("bs=1048576").arg("count=10")
             .stderr(Stdio::null())
             .status().expect("failed to execute `dd'");
         assert!(status.success());

